wallet, assumeutxo: Don't Assume m_chain_tx_count, Improve wallet RPC errors

fjahr opened the pull request:
A test that is added as part of #30455 uncovered this issue: The `GuessVerificationProgress` function is used during during descriptor import and relies on `m_chain_tx_count`. In #29370 an [`Assume` was added](https://github.com/bitcoin/bitcoin/pull/29370/commits/0fd915ee6bef63bb360ccc5c039a3c11676c38e3) expecting the `m_chaint_tx_count` to be set. However, as the test uncovered, `GuessVerificationProgress` is called with background sync blocks that have `m_chaint_tx_count = 0` when they have not been downloaded and processed yet.

The simple fix is to remove the `Assume`. Users should not be thrown off by the `Internal bug detected` error. The behavior of `importdescriptor` is kept consistent with the behavior for blocks missing due to pruning.

The test by alfonsoromanz is cherry-picked here to show that the [CI errors](https://cirrus-ci.com/task/5110045812195328?logs=ci#L2535) should be fixed by this change.

This PR also improves error messages returned by the `importdescriptors` and `rescanblockchain` RPCs. The error message now changes depending on the situation of the node, i.e. if pruning is happening or an assumutxo backgroundsync is active.

alfonsoromanz committed:
test, assumeutxo: import descriptors during background sync

alfonsoromanz reviewed (approved):
tACK 9be080abfa87543652c415af4e239a6ba4d2b2e1

The test passes locally, and I see that all the CI checks have passed, so it looks good to me.

It took me some time today to get familiar with the CMake-based build system, but I managed to get it working.

<img width="982" alt="Screenshot 2024-09-16 at 16 05 09" src="https://github.com/user-attachments/assets/bbe337ef-f789-49f0-a414-2661a17dbfb7">


Thanks!

mzumsande reviewed (commented):
> Whereas calling a rescan on the (missing) background blocks seems like a bug that should be avoided at the call site, no?

Well, the approach seems to be similar with pruning: 
`importdescriptors` tries to scan as many blocks as possible (no stopping after a failure due to missing block data), and then returns a failure / warns the user that the result may be incomplete. It could probably have been implentend differently, checking if we are missing blocks due to pruning and refuse to import the descriptors, but it doesn't do that.

If that is seen as a viable way for the pruning case, maybe it would be a bit inconsistent to just forbid it in the assumeutxo background sync case?! In which case just dealing with the progress estimation error would be ok?

mzumsande commented on src/validation.cpp:
```diff
@@ -5566,9 +5566,8 @@ double GuessVerificationProgress(const ChainTxData& data, const CBlockIndex *pin
-    if (!Assume(pindex->m_chain_tx_count > 0)) {
-        LogWarning("Internal bug detected: block %d has unset m_chain_tx_count (%s %s). Please report this issue here: %s\n",
-                   pindex->nHeight, PACKAGE_NAME, FormatFullVersion(), PACKAGE_BUGREPORT);
+    if (pindex->m_chain_tx_count == 0) {
```
I think this should only be possible to trigger with the rescan, the various RPCs are always be called with a block (usually the tip) that is guaranteed to have `m_chain_tx_count` set.

fjahr replied:
And aside from only hitting this in the context of rescan + blocks missing this is just about showing the progress and not the actual functionality the user seeks. So it didn't feel important enough for me to use another category mainly because of that. Unless there are strong opinions against it I will keep this as is.

mzumsande commented:
Though one could also make a case for not allowing importdescriptor/rescan during background sync, with the reasoning "just wait a little bit longer" that doesn't really apply to the pruning case. Would be interesting to know if this has ever been discussed for assumeutxo.

fjahr committed:
rpc: Include assumeutxo as a failure reason of rescanblockchain

fjahr commented:
> Can you explain this a bit better?

Yeah, my statement was wrong. I think I managed to confuse myself with some wrongfully placed debug logging. I have fixed the description.

> > Well, the approach seems to be similar with pruning:
> 
> Sure, if that is the case, the pull request description should mention it. Also, the error message could probably be updated, because it only mentions pruning, no AU.

Updated the error message and realized that he same issue exists in `rescanblockchain`, too. So fixed it there as well.

> Though one could also make a case for not allowing importdescriptor/rescan during background sync if non-downloaded blocks are affected, with the reasoning "just wait a little bit longer" that doesn't really apply to the pruning case. Would be interesting to know if this has ever been discussed for assumeutxo.

I had a similar thoughts when I looked into this earlier. But I think going that route might either break workflows with `importdescriptors` or it's a pretty invasive change. The RPC takes in multiple descriptors (requests) with timestamps for each and then rescans blocks starting from around the lowest timestamp. The simplest approach would be to check all the blocks from the lowest timestamp and abort the whole RPC if any of them don't have block data available for any block since the timestamp. Or we could exclude the offending requests and only import the remaining requests, starting rescan from the lowest timestamp with available block data. Both of these approaches seem pretty invasive and may lead to unexpected outcomes for users. I think people may undershoot the timestamp for example. We could then help them by returning the lowest timestamp that works (the snapshot height one) with the error but then we basically recreate the current behavior with extra steps. The ideal solution would be that the wallet rescans the missing blocks as they come in but that's probably a much bigger change, and also kind of tricky to communicate to the user.

Also, while "just wait a little bit longer" might be a good answer, we could also say "just run another rescan later" and that may ultimately be a better experience for the users.

I don't have a strong opinion here but it seems like these ideas may better be scoped out in a separate issue first and then potentially be implemented as a follow-up.

fjahr commented:
@achow101 do you have thoughts on descriptor import/rescan behavior with missing blocks because of ongoing assumeutxo background sync?

fjahr committed:
interfaces: Add helper function for wallet on pruning

fjahr committed:
validation: Don't assume m_chain_tx_count in GuessVerificationProgress

In the context of an a descriptor import during assumeutxo background sync, the progress can not be estimated due to m_chain_tx_count being set to 0.

fjahr committed:
rpc: Improve importdescriptor RPC error messages

Particularly add more details in the case of pruning or assumeutxo.

fjahr commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -909,9 +909,9 @@ RPCHelpMan rescanblockchain()
+        // We can't rescan unavailable blocks, stop and throw an error
         if (!pwallet->chain().hasBlocks(pwallet->GetLastBlockHash(), start_height, stop_height)) {
-            throw JSONRPCError(RPC_MISC_ERROR, "Can't rescan beyond pruned data. Use RPC call getblockchaininfo to determine your pruned height.");
+            throw JSONRPCError(RPC_MISC_ERROR, "Failed to rescan unavailable blocks. This may be caused by pruning or an in-progress assumeutxo background sync. Use RPC call getblockchaininfo to determine your pruned height or check your logs for assumeutxo progress.");
```
Improving the error message is reasonably easy but I didn't see a way without expanding the chain interface. I have implemented what I think is a step in the right direction but not a full solution to every edge case in order to keep the scope of this PR reasonable. Looking for feedback if people still think this is right approach and if I should expand it further or if adding more is going too far.

I think the assumeutxo + pruning use-case is pretty hard to deal with accurately for these wallet use-cases, especially if the pruning target is small. I am not sure if we can really give users a perfect answer in that case. Users could request a window for rescan that is larger than the prune target so I think pruning takes precedent because it requires a reindex to "fix" while AU just takes some time and resolves itself. Also, in the case of au+pruning a block could be not available because it wasn't synced yet during the first time the user tries to import a descriptor. But then the second time the user tries the same block that wasn't synced yet before could now be pruned already, assuming a small prune target. I mean to say, a perfect solution requires changing how the RPC works and I think that's beyond the scope of this PR. I could think about it a bit more and open an issue on it though if that is desirable.

fjahr commented:
Improved the error messages for `importdescriptors` and `rescanblockchain` to better match the reality (usage of pruning and assumeutxo).

fjahr commented on src/node/interfaces.cpp:
```diff
@@ -769,6 +770,16 @@ class ChainImpl : public Chain
     }
+    std::optional<int> getPruneHeight() override
+    {
+        LOCK(::cs_main);
```
done

fjahr replied:
fixed

fjahr commented:
Rebased and addressed nit from @maflcko

mzumsande reviewed (commented):
Concept ACK

fjahr commented on src/interfaces/chain.h:
```diff
@@ -289,6 +289,12 @@ class Chain
     virtual bool havePruned() = 0;
 
+    //! Get the current prune height.
+    virtual std::optional<int> getPruneHeight() = 0;
```
`hasBlocks()` is already used one line above where I use `getPruneHeight()`. The problem is it doesn't tell us if we have been pruning or not.

Is there a concerns with exposing `GetPruneHeight()` explicitly? I think the function could be moved outside of RPC code, it just wasn't needed anywhere else so far.

mzumsande replied:
Makes sense, I got confused there!
It just seemed a bit unusual, plus we generally use `CHECK_NONFATAL` instead of `assert` in rpc code, which may not be the ideal choice for possible future interface users (though it is currently just used in another rpc call, so it's not an actual problem).
But I'm not an expert on the design of our interfaces, so I don't know if moving it somewhere else is necessary.

Tiffanywaters92 replied:
So can I ask the worst question lol what is it that's wrong or needed with bitcoin with all of this? Simplest like for dummies kinda answer

fjahr commented on src/interfaces/chain.h:
```diff
@@ -289,6 +289,12 @@ class Chain
+    virtual std::optional<int> getPruneHeight() = 0;
+
+    //! Check if a snapshot chainstate exists.
+    virtual bool haveSnapshotChainstate() = 0;
```
Yeah, this works, I replaced it. Thanks!

fjahr commented:
Addressed feedback from @mzumsande

mzumsande reviewed (commented):
Code Review ACK 9d2d9f7ce29636f08322df70cf6abec8e0ca3727

furszy commented on src/validation.cpp:
```diff
@@ -5623,9 +5623,8 @@ double ChainstateManager::GuessVerificationProgress(const CBlockIndex* pindex) c
-        LogWarning("Internal bug detected: block %d has unset m_chain_tx_count (%s %s). Please report this issue here: %s\n",
-                   pindex->nHeight, CLIENT_NAME, FormatFullVersion(), CLIENT_BUGREPORT);
+    if (pindex->m_chain_tx_count == 0) {
+        LogDebug(BCLog::VALIDATION, "Block %d has unset m_chain_tx_count. Unable to estimate verification progress.\n", pindex->nHeight);
```
tiny nit: the `validation` category is a bit odd in the wallet scan context or GUI updates.

furszy reviewed (commented):
Code review ACK 9d2d9f7ce29636f08322df70cf6abec8e0ca3727

achow101 commented:
ACK 9d2d9f7ce29636f08322df70cf6abec8e0ca3727
